// m.config/lib/config.js

module.exports = {

  nodejump: {
    "0": "http://dev.nodeapi.ydl.com",
    "1": "http://dev.nodeapi.ydl.com",
    "2": "http://dev.nodeapi.ydl.com",
    "7": "https://testnodeapi.ydl.com",
    "9": "https://nodeapi.ydl.com",
  },

  dataBase: {
    G: {
      env: Config.env,
      time: Config.time,
      dist: Config.dist,
      webServer: Config.webServer,
      jslist: {},
    },

    meta: {
      title: "酷客村",
      keywords: "酷客村",
      description: "酷客村",
    },

    wechat(req){
      return {
        "title": "酷客村",
        "desc": "酷客村",
        "cover": "/src/images/global/wechat-share.png",
      }
    },

    preloadedState: {},
  },

	headers(res){
		res.set('Content-type', 'application/json');
		res.set('Access-Control-Allow-Origin', '*');
		res.set("Access-Control-Allow-Headers", 'X-Requested-With');
	},

	requestHeader: {
		headers: {

		}
	},

	timeout(res){

    res.setTimeout(30000, () => {
      console.log("响应超时");
      res.json(Object.assign({}, this.dataJson, {"code": "-1", "msg": "响应超时"}));
      return;
    });
	},

	dataJson: {
		code: 0,
		msg: "success",
		data: [],
  },
}
