// m.logs/lib/logs.js

const log4js = require('log4js');
const path = require('path');

module.exports = {

	init: function() {

    log4js.configure({

      "appenders": {

        "debug": {
          "type": 'file',
          "filename": path.join(Config.rootLogsProject, "/")+"debug.log",
          "maxLogSize": 3048000,
          "numBackups": 3,
          "level": "debug",
        },
        "trace": {
          "type": 'file',
          "filename": path.join(Config.rootLogsProject, "/")+"trace.log",
          "maxLogSize": 30480000,
          "numBackups": 1,
          "level": "trace",
        },
        "warn": {
          "type": 'file',
          "filename": path.join(Config.rootLogsProject, "/")+"warn.log",
          "maxLogSize": 30480000,
          "numBackups": 1,
          "level": "warn",
        },
        "info": {
          "type": "file",
          "filename": path.join(Config.rootLogsProject, "/")+"info.log",
          "maxLogSize": 30480000,
          "numBackups": 1,
          "level": "info",
        },
        "error": {
          "type": "file",
          "filename": path.join(Config.rootLogsProject, "/")+"error.log",
          "maxLogSize": 30480000,
          "numBackups": 1,
          "level": "error",
        },
      },

      "categories": {
        "info": {
          "appenders": ["info"],
          "level": Config.logs
        },
        "debug": {
          "appenders": ["debug"],
          "level": Config.logs
        },
        "error": {
          "appenders": ["error"],
          "level": Config.logs
        },
        "trace": {
          "appenders": ["trace"],
          "level": Config.logs
        },
        "warn": {
          "appenders": ["warn"],
          "level": Config.logs
        },
        "default": {
          "appenders": ["trace", "debug", "info", "warn", "error"],
          "level": "all"
        }
      },
    });

    return {
      debug(msg){
        log4js.getLogger('debug').debug("debug >>>>>>>>>>>>>>>>>>>>>>>");
        log4js.getLogger('debug').debug(msg || "");
      },
      error(msg){
        log4js.getLogger('error').error("error >>>>>>>>>>>>>>>>>>>>>>>");
        log4js.getLogger('error').error(msg || "");
      },
      info(msg){
        log4js.getLogger('info').info("info >>>>>>>>>>>>>>>>>>>>>>>");
        log4js.getLogger('info').info(msg || "");
      },
      warn(msg){
        log4js.getLogger('warn').warn("warn >>>>>>>>>>>>>>>>>>>>>>>");
        log4js.getLogger('warn').warn(msg || "");
      },
      trace(msg){
        log4js.getLogger('trace').trace("trace >>>>>>>>>>>>>>>>>>>>>>>");
        log4js.getLogger('trace').trace(msg || "");
      },
    }
	}
}
